### ISSUE TITLE
extraQueryParams/extraTokenParams/includeIdTokenInSilentRenew or SigninSilentArgs not respected when Refresh token is present in UserManager.signinSilent()

### ISSUE THREAD

**Issue by sachinsatav** [authts/oidc-client-ts#916]
So I need to pass some query params & access_token/id_token to my custom OAuth server token endpoint while refreshing token however configuration for extraQueryParams/extraTokenParams/includeIdTokenInSilentRenew or SigninSilentArgs is not taken into consideration when there is a refresh token for user.

https://github.com/authts/oidc-client-ts/blob/b4fe9078dd28533c9478c0472c06ee6fdd62457b/src/UserManager.ts#L255-L268

As per line 263, if there is a refresh token then no args are passed down on line 266. 

I need to pass my existing access token in Authorization header. Is there any way to solve this issue please? Thanks.

**Comment by pamapa** [comment]
The options extraQueryParams/extraTokenParams are used in the non refresh token case: using iframe. Can you paste the documentation of your IDP to refresh the token, that we have a chance to understand you case.
Best would be to copy/paste this library into your application and make working modifications and then post does modifications here...

**Comment by abennettHS** [comment]
I have this same issue with trying to support Azure AD B2C. It requires I pass through a `redirect_uri` despite not actually using it.

There appears to be a few issues open here that are related to the same issue of the params not being passed through for this refresh flow. Unsure how the library wants to approach this, but happy to help solve this issue.

These are my current throughts:

- Could try to use existing settings, such as extraQueryParams etc, but they are used elsewhere and may not be correct
- Could pass through, but that seems counter to design
- Pass through would make the interface in [react-oidc-context](https://github.com/authts/react-oidc-context) more accurate, as right now its misleading
- Could create a new setting, extraRefreshTokenParams or something similar

For anyone else who comes across this issue, specifically for Azure AD, here is my current work around:

```tsx
class AzureOidcClient extends OidcClient {
  constructor(settings: OidcClientSettings) {
    super(settings);
  }
 
  // copy pasted from the current repo
  async useRefreshToken({
    state,
    timeoutInSeconds,
  }: UseRefreshTokenArgs): Promise<SigninResponse> {
    const logger = this._logger.create("useAzureRefreshToken");

    // https://github.com/authts/oidc-client-ts/issues/695
    // In some cases (e.g. AzureAD), not all granted scopes are allowed on token refresh requests.
    // Therefore, we filter all granted scopes by a list of allowable scopes.
    let scope;
    if (this.settings.refreshTokenAllowedScope === undefined) {
      scope = state.scope;
    } else {
      const allowableScopes = this.settings.refreshTokenAllowedScope.split(" ");
      const providedScopes = state.scope?.split(" ") || [];

      scope = providedScopes.filter(s => allowableScopes.includes(s)).join(" ");
    }

    // desired redirect uri
    const redirect_uri = "urn:ietf:wg:oauth:2.0:oob";
    const result = await this._tokenClient.exchangeRefreshToken({
      refresh_token: state.refresh_token,
      // provide the (possible filtered) scope list
      scope,
      timeoutInSeconds,
      // @ts-ignore Redirect uri is not part of the base spec, but is required by Azure AD
      redirect_uri,
    });

    const response = new SigninResponse(new URLSearchParams());
    Object.assign(response, result);
    logger.debug("validating response", response);
    await this._validator.validateRefreshResponse(response, {
      ...state,
      // overide the scope in the state handed over to the validator
      // so it can set the granted scope to the requested scope in case none is included in the response
      scope,
    });
    return response;
  }
}

class AzureUserManager extends UserManager {
  constructor(config: UserManagerSettings) {
    super(config);
    // @ts-ignore Use our custom instance
    this._client = new AzureOidcClient(config);
  }
}
```

**Comment by pamapa** [comment]
fixed in [v3.0.0-rc.0](https://github.com/authts/oidc-client-ts/releases/tag/v3.0.0-rc.0)
